streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -1104,16 +1116,125 @@ public void cleanUp() {
      * {@code key} of this application or {@link StreamsMetadata#NOT_AVAILABLE} if Kafka Streams is (re-)initializing
+     * @deprecated Use {@link #queryMetadataForKey(String, Object, StreamPartitioner)} instead.
      */
+    @Deprecated
     public <K> StreamsMetadata metadataForKey(final String storeName,
                                               final K key,
                                               final StreamPartitioner<? super K, ?> partitioner) {
-        validateIsRunning();
+        validateIsRunningOrRebalancing();
         return streamsMetadataState.getMetadataWithKey(storeName, key, partitioner);
     }
 
+    /**
+     * Finds the metadata containing the active hosts and standby hosts where the key being queried would reside.
+     *
+     * @param storeName     the {@code storeName} to find metadata for
+     * @param key           the key to find metadata for
+     * @param keySerializer serializer for the key
+     * @param <K>           key type
+     * Returns {@link KeyQueryMetadata} containing all metadata about hosting the given key for the given store.
+     */
+    public <K> KeyQueryMetadata queryMetadataForKey(final String storeName,
+                                                    final K key,
+                                                    final Serializer<K> keySerializer) {
+        validateIsRunningOrRebalancing();
+        return streamsMetadataState.getKeyQueryMetadataForKey(storeName, key, keySerializer);
+    }
+
+    /**
+     * Finds the metadata containing the active hosts and standby hosts where the key being queried would reside.
+     *
+     * @param storeName     the {@code storeName} to find metadata for
+     * @param key           the key to find metadata for
+     * @param partitioner the partitioner to be use to locate the host for the key
+     * @param <K>           key type
+     * Returns {@link KeyQueryMetadata} containing all metadata about hosting the given key for the given store, using the
+     * the supplied partitioner
+     */
+    public <K> KeyQueryMetadata queryMetadataForKey(final String storeName,
+                                                    final K key,
+                                                    final StreamPartitioner<? super K, ?> partitioner) {
+        validateIsRunningOrRebalancing();
+        return streamsMetadataState.getKeyQueryMetadataForKey(storeName, key, partitioner);
+    }
+
+    /**
+     * Returns {@link LagInfo}, for all store partitions (active or standby) local to this Streams instance. Note that the
+     * values returned are just estimates and meant to be used for making soft decisions on whether the data in the store
+     * partition is fresh enough for querying.
+     *
+     * @return map of store names to another map of partition to {@link LagInfo}s
+     */
+    public Map<String, Map<Integer, LagInfo>> allLocalStorePartitionLags() {
```

Comment:
@vinothchandar , I took a closer look at the computations in this method. Here's my recommendation:

```java
    public Map<String, Map<Integer, LagInfo>> allLocalStorePartitionLags() {
        final long latestSentinel = -2L;
        final Map<String, Map<Integer, LagInfo>> localStorePartitionLags = new TreeMap<>();

        final Collection<TopicPartition> allPartitions = new LinkedList<>();
        final Map<TopicPartition, Long> changelogPositions = new HashMap<>();

        // Obtain the current positions, of all the active-restoring and standby tasks
        for (final StreamThread streamThread : threads) {
            for (final StandbyTask standbyTask : streamThread.allStandbyTasks()) {
                allPartitions.addAll(standbyTask.changelogPartitions());
                changelogPositions.putAll(standbyTask.restoredOffsets());
            }

            final Set<TaskId> restoringTaskIds = streamThread.restoringTaskIds();
            for (final StreamTask activeTask : streamThread.allStreamsTasks()) {
                final Collection<TopicPartition> taskChangelogPartitions = activeTask.changelogPartitions();
                allPartitions.addAll(taskChangelogPartitions);

                final boolean isRestoring = restoringTaskIds.contains(activeTask.id());
                final Map<TopicPartition, Long> restoredOffsets = activeTask.restoredOffsets();
                for (final TopicPartition topicPartition : taskChangelogPartitions) {
                    if (isRestoring && restoredOffsets.containsKey(topicPartition)) {
                        changelogPositions.put(topicPartition, restoredOffsets.get(topicPartition));
                    } else {
                        changelogPositions.put(topicPartition, latestSentinel);
                    }
                }
            }
        }

        log.debug("Current changelog positions: {}", changelogPositions);
        final Map<TopicPartition, ListOffsetsResultInfo> allEndOffsets;
        try {
            allEndOffsets = adminClient.listOffsets(
                allPartitions.stream()
                             .collect(Collectors.toMap(Function.identity(), tp -> OffsetSpec.latest()))
            ).all().get();
        } catch (final RuntimeException | InterruptedException | ExecutionException e) {
            throw new StreamsException("Unable to obtain end offsets from kafka", e);
        }
        log.debug("Current end offsets :{}", allEndOffsets);
        for (final Map.Entry<TopicPartition, ListOffsetsResultInfo> entry : allEndOffsets.entrySet()) {
            // Avoiding an extra admin API lookup by computing lags for not-yet-started restorations
            // from zero instead of the real "earliest offset" for the changelog.
            // This will yield the correct relative order of lagginess for the tasks in the cluster,
            // but it is an over-estimate of how much work remains to restore the task from scratch.
            final long earliestOffset = 0L;
            final long changelogPosition = changelogPositions.getOrDefault(entry.getKey(), earliestOffset);
            final long latestOffset = entry.getValue().offset();
            final LagInfo lagInfo = new LagInfo(changelogPosition == latestSentinel ? latestOffset : changelogPosition, latestOffset);
            final String storeName = streamsMetadataState.getStoreForChangelogTopic(entry.getKey().topic());
            localStorePartitionLags.computeIfAbsent(storeName, ignored -> new TreeMap<>())
                                   .put(entry.getKey().partition(), lagInfo);
        }

        return Collections.unmodifiableMap(localStorePartitionLags);
    }
```

Reply:
@vvcephei Trying to diff and understand what the changes are.. Will respond in a bit

Reply:
Changes seem to be :
- using a `TreeMap`,  (I guess for determinisitic ordering?)
- Use plain for loops
- Use `standbyTask.restoredOffsets()` instead of `standbyTask.checkpointedOffsets()`
- Changing of sentinel.. 

Don't know if I follow why the following is needed
```
            final long changelogPosition = changelogPositions.getOrDefault(entry.getKey(), earliestOffset);
```
To handle the case, where is a changelog partition for standby without any `restoredOffsets`? Could you give me an example?

Reply:
We plan to use `ProcessStateManager#standbyRestoredOffsets` for this.. We agreed `restoredOffsets` wont cut it for standbys